pkg/handlers/scripts/install.sh

```diff
@@ -1,116 +1,353 @@
+	command -v cut &> /dev/null || fail "cut not installed"
+	command -v du &> /dev/null || fail "du not installed"
+
+	# Check that the assets can be extracted.
+	case "${FTYPE}" in
+		".gz")
+			command -v gzip &>/dev/null || fail "gzip is not installed"
+			;;
+		".tar.gz")
+			command -v tar &>/dev/null || fail "tar is not installed"
+			;;
+		".zip")
+			command -v unzip &>/dev/null || fail "zip is not installed"
+			;;
+		"")
+			;;
+		*)
+			fail "unsupported file type ${FTYPE}"
+	esac
+
+	# Check if the install directory needs to be prompted for and exists.
+	[[ ! -z "${OUT_DIR:+x}" ]] || prompt_install_dir
+
+	if [[ ! -d "${OUT_DIR}" ]]; then
+		if [[ ! -z "${USE_SUDO+x}" ]]; then
+			sudo mkdir -p "${OUT_DIR}" &> /dev/null || true
+		else
+			mkdir -p "${OUT_DIR}" &> /dev/null || true
+		fi
+	fi
+
+	if [[ ! -w "${OUT_DIR}" && -z "${USE_SUDO+x}" ]]; then
+		echo ""
+		echo "The installation directory ${OUT_DIR} is not writeable by this user, and installation has failed."
+		echo ""
+		echo "To fix this, do one of the following:"
+		echo "  * Set the environment variable REPL_INSTALL_PATH to a directory in the PATH that can"
+		echo "    be written to and re-run this script"
+		echo "  * Set the environment variable REPL_USE_SUDO to any value and re-run this script. Keep"
+		echo "    in mind this script will block waiting on sudo."
+		echo "  * Re-run this script with the -i or --install flag set to a directory in the PATH that"
+		echo "    can be written to. 'curl http://kots.io | bash -s --install /new/path'"
+		echo "  * Re-run this script with the -s or --sudo flag. Keep in mind that this script will"
+		echo "    block waiting on sudo. 'curl http://kots.io | bash -s --sudo'"
+		echo "  * Re-run this script with sudo. 'curl http://kots.io | sudo bash'"
+		echo ""
+		fail "cannot write to the installation directory ${OUT_DIR}"
 	fi
-	#search subtree largest file (bin)
+}
+
+function install {
+	echo "Downloading ${USER}/${PROG} ${RELEASE} (${URL})..."
+
+	# Download and extract the binary to the temporary directory.
+	pushd $TMP_DIR &> /dev/null
+
+	local get_opts=("${INSECURE:+--insecure}" "--fail" "-#" "-L")
+
+	case "${FTYPE}" in
+		".gz")
+			curl "${get_opts[@]}" "${URL}" | gzip -d - > "${PROG}" || fail "download and extraction failed"
+			;;
+		".tar.gz")
+			curl "${get_opts[@]}" "${URL}" | tar xzf - > "${PROG}" || fail "download and extraction failed"
+			;;
+		".zip")
+			local tmp_file=$(basename $URL)
+		  curl "${get_opts[@]}" "${URL}" > "${tmp_file}" && unzip -o -qq "${tmp_file}" || fail "download and extraction failed"
+			rm "${tmp_file}"
+			;;
+		"")
+		  curl "${get_opts[@]}" "${URL}" > "{{ .Program }}_${OS}_${ARCH}" || fail "download failed"
+			;;
+		*)
+			fail "unknown file type ${FTYPE}"
+	esac
+
+	echo "Installing to ${OUT_DIR}"
+
+	# BUG: this will fail on a payload with unrelated files larger than the target binary.
+	# TODO: will there ever be unrelated files in the payload? Why not grab the _only_ file?
```

Comment:
This project was originally built for arbitrary binaries, we just forked it in order to add the `kubectl-` prefix to binaries (as required for kubectl plugins). So the original designers could not make that assumption.